fractions/microprofile/microprofile-jwt/src/main/java/org/wildfly/swarm/microprofile/jwtauth/deployment/auth/JWTAuthMechanism.java

```diff
@@ -64,54 +65,74 @@ public JWTAuthMechanism(JWTAuthContextInfo authContextInfo) {
+                // Install the JWT principal as the caller
+                Account account = identityManager.verify(credential.getName(), credential);
+                if (account != null) {
+                    JsonWebToken jwtPrincipal = (JsonWebToken) account.getPrincipal();
+                    securityContext.authenticationComplete(account, "MP-JWT", false);
+                    // Workaround authenticated JWTPrincipal not being installed as user principal
+                    // https://issues.jboss.org/browse/WFLY-9212
+                    org.jboss.security.SecurityContext jbSC = SecurityContextAssociation.getSecurityContext();
+                    Subject subject = jbSC.getUtil().getSubject();
+                    jbSC.getUtil().createSubjectInfo(jwtPrincipal, jwtToken, subject);
+                    RoleGroup roles = extract(subject);
+                    jbSC.getUtil().setRoles(roles);
+                    UndertowLogger.SECURITY_LOGGER.debugf("Authenticated caller(%s) for path(%s) with roles: %s",
+                            credential.getName(), exchange.getRequestPath(), account.getRoles());
+                    return AuthenticationMechanismOutcome.AUTHENTICATED;
+                } else {
+                    UndertowLogger.SECURITY_LOGGER.info("Failed to authenticate JWT bearer token");
+                    return AuthenticationMechanismOutcome.NOT_AUTHENTICATED;
                 }
+            } catch (Exception e) {
+                UndertowLogger.SECURITY_LOGGER.infof(e, "Failed to validate JWT bearer token");
+                return AuthenticationMechanismOutcome.NOT_AUTHENTICATED;
             }
         }
 
         // No suitable header has been found in this request,
         return AuthenticationMechanismOutcome.NOT_ATTEMPTED;
     }
 
+    private String getJwtToken(HttpServerExchange exchange) {
+        String bearerToken = null;
+        if (AUTHORIZATION.toString().equals(authContextInfo.getTokenHeader())) {
+            String authScheme = exchange.getRequestHeaders().getFirst(authContextInfo.getTokenHeader());
+            if (authScheme != null && authScheme.toLowerCase(Locale.ENGLISH).startsWith("bearer ")) {
+                bearerToken = authScheme.substring(7);
+            }
+        } else if (COOKIE.toString().equals(authContextInfo.getTokenHeader())
+            && authContextInfo.getTokenCookie() != null) {
+            List<String> cookieValues = exchange.getRequestHeaders().get(authContextInfo.getTokenHeader());
+            if (cookieValues != null) {
+                for (String cookieValue : cookieValues) {
+                    for (String cookie : cookieValue.split(";")) {
+                        if (cookie.startsWith(authContextInfo.getTokenCookie())) {
+                            String[] parts = cookie.split("=");
+                            if (parts.length == 2) {
+                                bearerToken = parts[1].trim();
```

Comment:
I'd very much prefer to not parse cookies here. There's a proper cookie parser in Undertow, let's use that.

Reply:
This code is straightforward meaning that, as far as the goal here is only capture a 'somecookie=token' pair which is not expected to contain anything else. We do not need a mechanism to capture all the cookies and make sure all is good even if no value is available or is some cookie has some extra properties. So is it worth bringing in the Undertow parser ? And if you feel yes, can you point to some example please ?

Reply:
Well, Undertow is already there, if I understand the code correctly, so it wouldn't bring any more dependencies.

Actually now that I'm thinking about it, Undertow has probably already parsed the cookies, so it should be a matter of looking up the correct one from `exchange.getRequestCookies()`, no?

Reply:
Let me check :-)

Reply:
Worked very well, much simpler now